File: algorithm/力扣热题100/19_删除链表的倒数第 N 个结点.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
    const dummy = { next: head }
    let slow = dummy, fast = dummy;
    n = n + 1
    while (n) {
        fast = fast.next;
        n--
    }
    while (fast) {
        fast = fast.next;
        slow = slow.next
    }
    slow.next = slow.next.next
    return dummy.next
};

File: algorithm/力扣热题100/19_删除链表的倒数第 N 个结点.go
package main

type ListNode2 struct {
	Val  int
	Next *ListNode2
}

func removeNthFromEnd(head *ListNode2, n int) *ListNode2 {
	dummy := &ListNode2{
		Val:  0,
		Next: nil,
	}
	fast := dummy
	slow := dummy
	n++
	for n > 0 && fast != nil {
		n--
		fast = fast.Next
	}
	for fast != nil {
		fast = fast.Next
		slow = slow.Next
	}
	slow.Next = slow.Next.Next
	return dummy.Next
}
